src/lava/magma/core/model/py/ports.py

```diff
@@ -20,7 +21,22 @@ def csp_ports(self) -> ty.List[AbstractCspPort]:
         pass
 
 
-class PyInPort(AbstractPyPort):
+class AbstractIOPyPort(AbstractPyPort):
+    def __init__(self,
+                 csp_ports: ty.List[AbstractCspPort],
+                 process_model: AbstractProcessModel,
+                 shape: ty.Tuple[int, ...] = tuple(),
```

Comment:
Not sure if default arguments should be provided in AbstractClass. They should be provided and documented in most concrete class. Imagine the concrete class allowing non-kwargs arguments and then users will get errors for not initializing some but not other arguments. How confusing would that be.

Reply:
I'm not against the arguments as such. I'm questioning the usefulness of default argument assignments in abstract classes.

In particular, what's the purpose of an empty tuple for 'shape'? That does not seem to be a reasonable value anyways. This should probably be a required argument without default. If there's no shape -> Error.

Perhaps a concrete class has a good way of setting default arguments. But it seems an empty tuple is meaningless in a AbstractClass. If not provided, it will later fail somewhere in the weeks because of an empty tuple and it will not be clear why this error shows up at this point.

We should probably even add an assertion or raise an error if an empty tuple is passed...?

Reply:
Oh okay, I overlooked the default part. Removed the defaults.